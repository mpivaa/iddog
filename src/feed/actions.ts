import { get } from '@app/api';
import { getToken } from '@app/auth';
import { State } from '@app/store';
import { Error } from '@app/types';
import { Dispatch } from 'redux';

export interface FetchFeedPayload {
  category: string;
  list: string[];
}

export enum ActionType {
  REQUEST = 'FEED_REQUEST',
  SUCCESS = 'FEED_SUCCESS',
  FAILURE = 'FEED_FAILURE',
}

export const fetchFeedRequest = () => ({ type: ActionType.REQUEST });
export const fetchFeedSuccess = (data: FetchFeedPayload) =>
  ({ type: ActionType.SUCCESS, data });
export const fetchFeedError = (error: Error) =>
  ({ type: ActionType.FAILURE, error });

export function fetchFeed(category: string) {
  return async (dispatch: Dispatch, getState: () => State) => {
    dispatch(fetchFeedRequest());
    const token = getToken(getState());

    const { data, error } = await get<FetchFeedPayload>('/feed', {
      token: token!,
      params: {
        category,
      },
    });

    if (error) {
      return dispatch(fetchFeedError(error));
    } else {
      return dispatch(fetchFeedSuccess(data!));
    }
  };
}

export type Action =
  | { type: typeof ActionType.REQUEST }
  | { type: typeof ActionType.SUCCESS, data: FetchFeedPayload }
  | { type: typeof ActionType.FAILURE, error: Error };
