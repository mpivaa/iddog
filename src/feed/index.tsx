export { default as Feed } from './Feed';
export { default as reducer, State } from './reducer';
export * from './selectors';
export * from './actions';
