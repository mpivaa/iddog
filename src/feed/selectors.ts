import { getQuery } from '@app/router';
import { State } from '@app/store';

export const getDataList = (state: State) => state.feed.data.list;
export const getError = (state: State) => state.feed.error;
export const getCategory = (state: State, initialValue: string): string =>
  getQuery(state).category || initialValue;
export const getId = (state: State): string =>
  getQuery(state).id;
