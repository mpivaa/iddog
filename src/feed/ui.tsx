import * as React from 'react';
import { Link, LinkProps } from 'react-router-dom';
import styled from 'styled-components';

interface Props extends LinkProps {
  to: string;
  active?: boolean;
}

function Category({ to, ...rest }: Props) {
  return (
    <Link
      {...rest}
      to={{
        search: `?category=${to}`,
      }}
    />
  );
}

const StyledCategory = styled(Category)`
  color: ${(props) => props.active ? '#333' : '#aaa'};
  font-weight: ${(props) => props.active ? 'bold' : 'normal'};
  text-decoration: none;
  margin: 12px;

  &:hover {
    color: '#333';
    font-weight: ${(props) => props.active ? 'bold' : 'normal'};
  }
`;

const CloseButton = styled.div`
  position: absolute;
  top: 8px;
  right: 8px;
  font-size: 18px;
  color: transparent;
  text-shadow: 0 0 0 white;
  cursor: pointer;
`;

export { StyledCategory as Category, CloseButton };
