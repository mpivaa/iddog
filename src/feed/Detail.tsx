import * as React from 'react';
import { Container, LazyImage, LinearLayout, Overlay } from '@app/ui';
import { CloseButton } from '@app/feed/ui';

interface Props {
  id: string;
  goBack(): void;
}

export const Detail: React.SFC<Props> = ({ id, goBack }) => {
  return (
    <Overlay onClick={() => goBack()}>
      <Container size="large">
        <LinearLayout fill direction="column">
          <LinearLayout direction="row">
            <CloseButton>‚ùå</CloseButton>
            <LazyImage src={id} size="large" />
          </LinearLayout>
        </LinearLayout>
      </Container>
    </Overlay>
  );
};

export default Detail;
