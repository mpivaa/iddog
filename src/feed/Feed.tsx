import * as React from 'react';
import { goBack, push } from 'connected-react-router';
import { connect } from 'react-redux';
import { fetchFeed } from '@app/feed/actions';
import Detail from '@app/feed/Detail';
import { getCategory, getDataList, getError, getId } from '@app/feed/selectors';
import { Category } from '@app/feed/ui';
import { State as AppState } from '@app/store';
import { Error } from '@app/types';
import { Container, LazyImage, LinearLayout, Title } from '@app/ui';

interface Props {
  category: string;
  id: string;
  list: string[];
  error: Error | null;
  fetchFeed(category: string): void;
  promoteItem(src: string, category: string): void;
  goBack(): void;
}

export class Feed extends React.Component<Props> {
  public componentDidMount() {
    this.props.fetchFeed(this.props.category);
  }

  public componentDidUpdate(prevProps: Props) {
    if (this.props.category !== prevProps.category) {
      this.props.fetchFeed(this.props.category);
    }
  }

  public render() {
    const { list, category, promoteItem, id, goBack } = this.props;
    const categories: { [key: string]: string } = {
      husky: 'Husky',
      labrador: 'Labrador',
      hound: 'Hound',
      pug: 'Pug',
    };

    return (
      <Container size="large">
        <LinearLayout direction="column">
          <Title>The IDDOG</Title>
          <LinearLayout direction="row">
            {Object.keys(categories).map((key) => (
              <Category key={key} to={key} active={key === category}>
                {categories[key]}
              </Category>
            ))}
          </LinearLayout>
          <LinearLayout direction="row" wrap>
            {list.map((src) => (
              <LazyImage
                key={src}
                cover
                size="medium"
                src={src}
                onClick={() => promoteItem(src, category)}
              />
            ))}
          </LinearLayout>
        </LinearLayout>
        {id && <Detail id={id} goBack={goBack} />}
      </Container>
    );
  }
}

const mapState = (state: AppState) => ({
  list: getDataList(state),
  error: getError(state),
  category: getCategory(state, 'husky'),
  id: getId(state),
});

const mapDispatch = {
  fetchFeed,
  promoteItem(src: string, category: string) {
    return push({
      search: `?category=${category}&id=${src}`,
    });
  },
  goBack,
};

export default connect(mapState, mapDispatch)(Feed);
