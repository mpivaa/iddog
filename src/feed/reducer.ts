import { Action, ActionType, FetchFeedPayload } from '@app/feed/actions';
import { Error } from '@app/types';

export interface State {
  loading: boolean;
  data: FetchFeedPayload;
  error: Error | null;
}

const initialState: State = {
  loading: false,
  data: {
    category: '',
    list: [],
  },
  error: null,
};

function reducer(state = initialState, action: Action) {
  switch (action.type) {
    case ActionType.REQUEST: return {
      ...state,
      loading: true,
    };
    case ActionType.SUCCESS: return {
      ...state,
      loading: false,
      data: action.data,
      error: null,
    };
    case ActionType.FAILURE: return {
      ...state,
      loading: false,
      error: action.error,
    };
    default: return state;
  }
}

export default reducer;
