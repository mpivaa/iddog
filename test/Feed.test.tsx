import * as React from 'react';
import { mount } from 'enzyme';
import { Feed } from '@app/feed';
import Setup, { fetchMock, store } from '@test/setup';

const fixture = {
  husky: {
    category: 'husky',
    list: [
      'https://dog.ceo/api/img/husky/n02110185_10047.jpg',
      'https://dog.ceo/api/img/husky/n02110185_10116.jpg'
    ]
  },
  labrador: {
    category: 'labrador',
    list: [
      'https://dog.ceo/api/img/labrador/n02099712_1150.jpg',
      'https://dog.ceo/api/img/labrador/n02099712_1200.jpg',
      'https://dog.ceo/api/img/labrador/n02099712_1229.jpg'
    ]
  }
};

fetchMock.mock('end:/feed?category=husky', fixture.husky);
fetchMock.mock('end:/feed?category=labrador', fixture.labrador);

test('Feed renders correctly', (done) => {
  const component = mount(
    <Setup>
      <Feed />
    </Setup>
  );

  setTimeout(() => {
    const rendered = component.render();
    expect(rendered).toMatchSnapshot();
    expect(store.getState().feed).toMatchSnapshot();
    done();
  }, 500);
});

test('Set category', (done) => {
  const component = mount(
    <Setup>
      <Feed />
    </Setup>
  );

  expect(component.render()).toMatchSnapshot();

  component.find({ href: '/?category=labrador' })
    .simulate('click', { button: 0 });

  setTimeout(() => {
    expect(component.render()).toMatchSnapshot();
    expect(store.getState().feed).toMatchSnapshot();
    done();
  }, 500);
});

test('Show item', (done) => {
  const component = mount(
    <Setup>
      <Feed />
    </Setup>
  );

  setTimeout(() => {
    component.find('LazyImage').first().props().onClick!({} as any);
    expect(component.render()).toMatchSnapshot();
    expect(store.getState().router.location.search).toMatchSnapshot();
    done();
  }, 500);
});
