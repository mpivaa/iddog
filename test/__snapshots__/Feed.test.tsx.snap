// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed renders correctly 1`] = `
<div
  class="sc-bxivhb oUyRg"
>
  <div
    class="sc-htpNat jaBiNQ"
    direction="column"
  >
    <div
      class="sc-EHOje cYMhyC"
    >
      The IDDOG
    </div>
    <div
      class="sc-htpNat coRsnH"
      direction="row"
    >
      <a
        class="sc-dnqmqq bUwXkI"
        href="/?category=husky"
      >
        Husky
      </a>
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=labrador"
      >
        Labrador
      </a>
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=hound"
      >
        Hound
      </a>
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=pug"
      >
        Pug
      </a>
    </div>
    <div
      class="sc-htpNat IQKaj"
      direction="row"
    >
      <div
        class="sc-bZQynM SLkpq"
      >
        <div
          class="lazyload-placeholder"
          style="height: 270px;"
        />
      </div>
      <div
        class="sc-bZQynM SLkpq"
      >
        <div
          class="lazyload-placeholder"
          style="height: 270px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Feed renders correctly 2`] = `
Object {
  "data": Object {
    "category": "husky",
    "list": Array [
      "https://dog.ceo/api/img/husky/n02110185_10047.jpg",
      "https://dog.ceo/api/img/husky/n02110185_10116.jpg",
    ],
  },
  "error": null,
  "loading": false,
}
`;

exports[`Set category 1`] = `
<div
  class="sc-bxivhb oUyRg"
>
  <div
    class="sc-htpNat jaBiNQ"
    direction="column"
  >
    <div
      class="sc-EHOje cYMhyC"
    >
      The IDDOG
    </div>
    <div
      class="sc-htpNat coRsnH"
      direction="row"
    >
      <a
        class="sc-dnqmqq bUwXkI"
        href="/?category=husky"
      >
        Husky
      </a>
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=labrador"
      >
        Labrador
      </a>
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=hound"
      >
        Hound
      </a>
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=pug"
      >
        Pug
      </a>
    </div>
    <div
      class="sc-htpNat IQKaj"
      direction="row"
    >
      <div
        class="sc-bZQynM SLkpq"
      >
        <div
          class="lazyload-placeholder"
          style="height: 270px;"
        />
      </div>
      <div
        class="sc-bZQynM SLkpq"
      >
        <div
          class="lazyload-placeholder"
          style="height: 270px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Set category 2`] = `
<div
  class="sc-bxivhb oUyRg"
>
  <div
    class="sc-htpNat jaBiNQ"
    direction="column"
  >
    <div
      class="sc-EHOje cYMhyC"
    >
      The IDDOG
    </div>
    <div
      class="sc-htpNat coRsnH"
      direction="row"
    >
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=husky"
      >
        Husky
      </a>
      <a
        class="sc-dnqmqq bUwXkI"
        href="/?category=labrador"
      >
        Labrador
      </a>
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=hound"
      >
        Hound
      </a>
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=pug"
      >
        Pug
      </a>
    </div>
    <div
      class="sc-htpNat IQKaj"
      direction="row"
    >
      <div
        class="sc-bZQynM SLkpq"
      >
        <div
          class="lazyload-placeholder"
          style="height: 270px;"
        />
      </div>
      <div
        class="sc-bZQynM SLkpq"
      >
        <div
          class="lazyload-placeholder"
          style="height: 270px;"
        />
      </div>
      <div
        class="sc-bZQynM SLkpq"
      >
        <div
          class="lazyload-placeholder"
          style="height: 270px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Set category 3`] = `
Object {
  "data": Object {
    "category": "labrador",
    "list": Array [
      "https://dog.ceo/api/img/labrador/n02099712_1150.jpg",
      "https://dog.ceo/api/img/labrador/n02099712_1200.jpg",
      "https://dog.ceo/api/img/labrador/n02099712_1229.jpg",
    ],
  },
  "error": null,
  "loading": false,
}
`;

exports[`Show item 1`] = `
<div
  class="sc-bxivhb oUyRg"
>
  <div
    class="sc-htpNat jaBiNQ"
    direction="column"
  >
    <div
      class="sc-EHOje cYMhyC"
    >
      The IDDOG
    </div>
    <div
      class="sc-htpNat coRsnH"
      direction="row"
    >
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=husky"
      >
        Husky
      </a>
      <a
        class="sc-dnqmqq bUwXkI"
        href="/?category=labrador"
      >
        Labrador
      </a>
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=hound"
      >
        Hound
      </a>
      <a
        class="sc-dnqmqq ckvkZp"
        href="/?category=pug"
      >
        Pug
      </a>
    </div>
    <div
      class="sc-htpNat IQKaj"
      direction="row"
    >
      <div
        class="sc-bZQynM SLkpq"
      >
        <div
          class="lazyload-placeholder"
          style="height: 270px;"
        />
      </div>
      <div
        class="sc-bZQynM SLkpq"
      >
        <div
          class="lazyload-placeholder"
          style="height: 270px;"
        />
      </div>
      <div
        class="sc-bZQynM SLkpq"
      >
        <div
          class="lazyload-placeholder"
          style="height: 270px;"
        />
      </div>
    </div>
  </div>
  <div
    class="sc-htoDjs coBocp"
  >
    <div
      class="sc-bxivhb oUyRg"
    >
      <div
        class="sc-htpNat dSBIma"
        direction="column"
      >
        <div
          class="sc-htpNat coRsnH"
          direction="row"
        >
          <div
            class="sc-bZQynM dpxuNW"
          >
            <div
              class="lazyload-placeholder"
              style="height: 600px;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Show item 2`] = `"?category=labrador&id=https://dog.ceo/api/img/labrador/n02099712_1150.jpg"`;
